File: 0875-koko-eating-bananas/0875-koko-eating-bananas.rs
impl Solution {
    pub fn min_eating_speed(piles: Vec<i32>, h: i32) -> i32 {
        let mut left: i32 = 1;
        let mut right: i32 = *piles.iter().max().unwrap();
        
        while left < right {
            let mid: i32 = (left + right) / 2;
            let mut hours: i32 = 0;
            
            for pile in &piles {
                hours += (*pile as f64 / mid as f64).ceil() as i32
            }
            
            if hours <= h {
                right = mid
            } else {
                left = mid + 1
            }
        }
        return right
    }
}

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    auto bananEatTime(int speed, vector<int> &piles) -> long long {
        long long res = 0;
        for (auto &pile:piles) {
            res += ceil((double)pile/speed);
        }
        
        return res;
    }
    
    int minEatingSpeed(vector<int>& piles, int h) {
        long long l = 1,
            r = *std::max_element(piles.begin(),piles.end());
        long long res = r;
        while (l <= r) {
            long long m = l + (r - l)/2,
                t = bananEatTime(m, piles);
            
            if (t > h) {
                l = m + 1;
            } else /*if (t <= h)*/ {
                r = m - 1;
                res = std::min(res,m);
            }
        }
        
        return (int)res;
    }
};